### ISSUE TITLE
feat: Implement backward kernels for training support (CUDA autograd)

### ISSUE THREAD

**Issue by noahgift** [paiml/trueno#85]
## Summary

**trueno-gpu** currently provides forward-pass CUDA kernels (GEMM, softmax, attention, layer_norm, etc.) but **no backward-pass kernels**. This blocks all downstream training workloads in the PAIML stack.

**Requesting Project:** `entrenar` (Rust training library)
**Blocking Issue:** SPEC-FT-001 Phase 11 (CUDA Migration)
**Priority:** P0 - Critical Path Blocker

---

## Context: Why trueno-gpu?

Five-Whys analysis determined backward kernels belong in `trueno-gpu` because:

1. **Forward + Backward are paired** - mathematically coupled operations
2. **PTX builder already exists** - infrastructure is here
3. **Compute primitive layer** - lowest common denominator
4. **Clean dependency graph:**
   ```
   entrenar → trueno-gpu ✓
   aprender → trueno-gpu ✓
   realizar → trueno-gpu ✓
   ```
5. **DRY principle** - one optimized implementation, many consumers

---

### 1. GEMM Backward (`gemm_backward_cuda`)

**Mathematical Specification:**
```
Forward:  C = A @ B           where A: [M×K], B: [K×N], C: [M×N]
Backward: ∂L/∂A = ∂L/∂C @ Bᵀ  → [M×K]
          ∂L/∂B = Aᵀ @ ∂L/∂C  → [K×N]
```

**API:**
```rust
pub struct GemmBackwardKernel {
    m: u32,
    k: u32,
    n: u32,
}

impl GemmBackwardKernel {
    /// Returns (grad_a_kernel, grad_b_kernel) as separate PTX modules
    pub fn build_ptx(&self) -> (PtxKernel, PtxKernel);
}
```

**Falsifiable Prediction (P-GEMM-BACK-001):**
> For random matrices A[512×512], B[512×512], the analytical gradient from `gemm_backward_cuda` will match finite-difference numerical gradient within ε < 1e-3 for 99.9% of elements.

---

### 2. Softmax Backward (`softmax_backward_cuda`)

**Mathematical Specification:**
```
Forward:  y_i = exp(x_i) / Σ_j exp(x_j)
Backward: ∂L/∂x_i = y_i * (∂L/∂y_i - Σ_j(∂L/∂y_j * y_j))
        = y_i * (∂L/∂y_i - dot(∂L/∂y, y))
```

**Key Implementation Detail:** The backward pass requires the *output* of softmax (y), not the input (x). This is more numerically stable and avoids recomputation.

**API:**
```rust
pub struct SoftmaxBackwardKernel {
    size: u32,
}

impl SoftmaxBackwardKernel {
    /// grad_input = softmax_output * (grad_output - dot(grad_output, softmax_output))
    pub fn build_ptx(&self) -> PtxKernel;
}
```

**Falsifiable Prediction (P-SOFT-BACK-001):**
> For vectors of length 1024 with values in [-10, 10], analytical gradient matches finite-difference within ε < 1e-4.

**Falsifiable Prediction (P-SOFT-BACK-002):**
> Kernel handles edge case where one element dominates (softmax ≈ [0, 0, ..., 1, ..., 0]) without NaN/Inf.

---

### 3. LayerNorm Backward (`layer_norm_backward_cuda`)

**Mathematical Specification:**
```
Forward:  y = γ * (x - μ) / σ + β
          where μ = mean(x), σ = sqrt(var(x) + ε)

Backward: ∂L/∂γ = Σ_i (∂L/∂y_i * x̂_i)        where x̂ = (x - μ) / σ
          ∂L/∂β = Σ_i (∂L/∂y_i)
          ∂L/∂x = (1/σ) * (∂L/∂y - mean(∂L/∂y) - x̂ * mean(∂L/∂y * x̂)) * γ
```

**API:**
```rust
pub struct LayerNormBackwardKernel {
    hidden_size: u32,
    eps: f32,
}

impl LayerNormBackwardKernel {
    /// Returns gradients for (input, gamma, beta)
    pub fn build_ptx(&self) -> PtxKernel;
}
```

**Falsifiable Prediction (P-LN-BACK-001):**
> Gradient w.r.t. input matches finite-difference within ε < 1e-4 for hidden_size in [64, 128, 256, 512, 896, 1024, 2048, 4096].

**Falsifiable Prediction (P-LN-BACK-002):**
> Gradient w.r.t. gamma and beta are correct for learned affine parameters.

---

### 4. Flash Attention Backward (`flash_attention_backward_cuda`)

**Mathematical Specification:**
```
Forward:  O = softmax(Q @ Kᵀ / √d) @ V
          where Q: [N×d], K: [N×d], V: [N×d], O: [N×d]

Backward: (Complex - requires storing LSE from forward pass)
          ∂L/∂Q = ∂L/∂O @ Vᵀ @ ∂softmax/∂(QKᵀ) @ K / √d
          ∂L/∂K = similar...
          ∂L/∂V = softmax(QKᵀ/√d)ᵀ @ ∂L/∂O
```

**Key Implementation Detail:** Flash attention backward requires the log-sum-exp (LSE) values from the forward pass for numerical stability. The forward kernel must be modified to optionally output LSE.

**API:**
```rust
pub struct FlashAttentionBackwardKernel {
    seq_len: u32,
    head_dim: u32,
    causal: bool,
}

impl FlashAttentionBackwardKernel {
    /// Requires: grad_output, Q, K, V, output, lse (log-sum-exp from forward)
    /// Returns: (grad_Q, grad_K, grad_V)
    pub fn build_ptx(&self) -> PtxKernel;
}
```

**Falsifiable Prediction (P-ATTN-BACK-001):**
> For seq_len=128, head_dim=64, gradients match PyTorch's `F.scaled_dot_product_attention` backward within ε < 1e-3.

**Falsifiable Prediction (P-ATTN-BACK-002):**
> Causal masking is correctly applied in backward pass (gradients for masked positions are zero).

---

### 5. Activation Backward Kernels

**ReLU Backward:**
```
Forward:  y = max(0, x)
Backward: ∂L/∂x = ∂L/∂y * (x > 0 ? 1 : 0)
```

**GELU Backward:**
```
Forward:  y = 0.5 * x * (1 + tanh(√(2/π) * (x + 0.044715 * x³)))
Backward: ∂L/∂x = ∂L/∂y * gelu'(x)  // Derivative is complex but well-defined
```

**SiLU/Swish Backward:**
```
Forward:  y = x * sigmoid(x)
Backward: ∂L/∂x = ∂L/∂y * (sigmoid(x) + x * sigmoid(x) * (1 - sigmoid(x)))
        = ∂L/∂y * (y + sigmoid(x) * (1 - y))
```

**API:**
```rust
pub struct ReluBackwardKernel { size: u32 }
pub struct GeluBackwardKernel { size: u32 }
pub struct SiluBackwardKernel { size: u32 }
```

**Falsifiable Prediction (P-ACT-BACK-001):**
> All activation backward kernels match finite-difference within ε < 1e-5.

---

### 6. RMSNorm Backward (`rms_norm_backward_cuda`)

**Mathematical Specification:**
```
Forward:  y = x / rms(x) * γ    where rms(x) = √(mean(x²) + ε)
Backward: ∂L/∂γ = Σ_i (∂L/∂y_i * x_i / rms)
          ∂L/∂x = γ/rms * (∂L/∂y - x * mean(∂L/∂y * x) / (rms² + ε))
```

**Falsifiable Prediction (P-RMS-BACK-001):**
> Matches PyTorch's RMSNorm backward within ε < 1e-4.

---

## Performance Targets

| Kernel | Size | Target Latency | Baseline (PyTorch) |
|--------|------|----------------|-------------------|
| gemm_backward | 512×512 | <1ms | ~0.8ms |
| softmax_backward | 10K | <0.1ms | ~0.08ms |
| layer_norm_backward | 896×1024 | <0.2ms | ~0.15ms |
| flash_attn_backward | 2048×64 | <5ms | ~4ms |
| gelu_backward | 1M | <0.05ms | ~0.04ms |

**Meta-Prediction (P-PERF-001):**
> All backward kernels will be within 2× of PyTorch's CUDA implementations on RTX 4090.

---

## File Structure

```
trueno-gpu/src/kernels/
├── mod.rs                    # Add backward module
├── backward/
│   ├── mod.rs               # Re-exports
│   ├── gemm.rs              # GemmBackwardKernel
│   ├── softmax.rs           # SoftmaxBackwardKernel
│   ├── layer_norm.rs        # LayerNormBackwardKernel
│   ├── rms_norm.rs          # RmsNormBackwardKernel
│   ├── attention.rs         # FlashAttentionBackwardKernel
│   └── activations.rs       # ReLU, GELU, SiLU backward
```

---

## Testing Requirements

1. **Property-Based Tests:** 10,000+ cases per kernel with random inputs
2. **Finite-Difference Validation:** Numerical gradient comparison (ε=1e-5)
3. **Edge Cases:** 
   - Near-zero values (vanishing gradients)
   - Large values (overflow)
   - Single dominant element (softmax saturation)
4. **Barrier Safety:** All kernels must pass `validate_barrier_safety()` (PARITY-114)
5. **Determinism:** Same inputs → same outputs (for reproducibility)

---

## Acceptance Criteria

- [ ] All 6 kernel families implemented with PTX generation
- [ ] Property tests pass (10K+ cases each)
- [ ] Finite-difference validation passes (ε < 1e-4)
- [ ] Barrier safety validation passes
- [ ] Performance within 2× of PyTorch baseline
- [ ] Documentation with mathematical derivations
- [ ] Examples in `trueno-gpu/examples/backward_*.rs`

---

## References

- [FlashAttention-2 Paper](https://arxiv.org/abs/2307.08691) - Section 3.2 for backward algorithm
- [PyTorch Autograd](https://pytorch.org/docs/stable/notes/autograd.html) - Reference implementation
- [CUDA Math API](https://docs.nvidia.com/cuda/cuda-math-api/) - Intrinsics for exp, tanh, etc.

---

## Downstream Impact

**Unblocks:**
- `entrenar` SPEC-FT-001 Phase 11 (CUDA Migration)
- `entrenar` fine-tuning at 100× current throughput
- Future `aprender` neural network training
- Future `realizar` on-device fine-tuning

**Current Bottleneck (entrenar):**
```
nvidia-smi during training:
- GPU Utilization: 10%  ← Forward on GPU
- CPU Utilization: 100% ← Backward on CPU (ndarray)
- Token generation: ~1 tok/s (should be 100+)
```

With backward kernels: **100× speedup expected**.

---

*"A theory is scientific if and only if it is falsifiable."* — Karl Popper

All predictions above are falsifiable. If any fail, the implementation is incorrect.
